### DISCUSSION TITLE
Trying to mount nfs share into rootless container for user-data

### DISCUSSION THREAD

**Discussion by MagicTrevor** [containers/podman#12499] [General]
I could use some guidance in regards to rootless podman and NFS shares to store user data only, not to run container images from.

I am probably grossly misunderstanding this, but I have an NFS share mounted in the host and am trying to mount that directory as a volume into a container and receive the following error:
`Error: lsetxattr /mnt/mynfs/: operation not supported`

`-v /mnt/mynfs:/path/in/container:z`

**Comment by rhatdan** [comment]
Don't add the :z.  NFS directories sometimes do not support XATTRs.  I think policy should allow this to work without the relabeling.  NFS does support XATTRs but you need to configure the client and server to make it work.

https://www.thegeeksearch.com/how-to-configure-selinux-labeled-nfs-exports/

**Reply by MagicTrevor** [reply]
If I leave the :z option off then the container can only read... is that a matter of the uid/guid at that point?

**Reply by rhatdan** [reply]
Setenforce 0 to see if it is SELinux.
What does ls -l NFSFILES inside of the container show for ownership?
How does your user ordinarily access these files? if it is via supplimental groups then you need to look at

--group-add keep-groups

**Reply by MagicTrevor** [reply]
Setting `Setenforce 0` and removing the `:z` allows me to write to the nfs share now.
Output of `ls -l /test` (which is where I have mounted the nfs share for now):
![image](https://user-images.githubusercontent.com/27883110/144668422-bf0c1c84-6458-4c1b-ab90-0b1af50363ad.png)
Normally the container manages that and should own as the user www-data or 82.

**Reply by MagicTrevor** [reply]
The "root" user of the container `uid=0(root) gid=0(root)` can properly read-write files on the nfs mount. However the app in the container run as the www-data user `uid=82(www-data) gid=82(www-data) groups=82(www-data)`. Am I on track in that I need to somehow map the uid 82 to the host user as well?

**Reply by rhatdan** [reply]
What AVCs were you seeing?
Perhaps running with --security-opt label=disable would be better then setenforce 0.  Since only this contianer would be unconfined.

**Reply by MagicTrevor** [reply]
AVCs?

**Reply by rhatdan** [reply]
grep -i avc /var/log/audit/audit.log

**Reply by MagicTrevor** [reply]
The only thing currently is `type=AVC msg=audit(1639056325.606:171): avc:  denied  { name_bind } for  pid=1207 comm="rpcbind" src=62858 scontext=system_u:system_r:rpcbind_t:s0 tcontext=system_u:object_r:unreserved_port_t:s0 tclass=udp_socket permissive=0`.

It won't let me use `podman unshare` in order to try to change the directory ownership to the right user. On the server side it is mounted under the user running podman. `ls -al` within the container shows everything owned by `root:root` but the parent is owned by www-data.

**Reply by MagicTrevor** [reply]
I added the --security-opt label=disable line in instead of the setenforce 0.

EDIT: I'll add that www-data can read from the nfs mount and can create new files, but cannot modify existing files. I believe part of the container's startup is it tries to chmod the directory to the www-data user.

**Reply by rhatdan** [reply]
Great does not look like SELinux AVC messages about containers.  Does the label=disable allow it to work?

**Reply by MagicTrevor** [reply]
It is mounted inside the container and the container root user can read/write but the actual app user cannot (since it's owned by root)

**Reply by rhatdan** [reply]
You would need to chown the file to match the user of the app
-v /SOURCE:/DEST:U  might do what you need, this would tell Podman to chown the files to the correct permissions.  Only problem would be NFS server would likely not allow a rootless user to chown the files.

Another option would be to setup group read/write on the files and then leak the group access into the container.

--group-add=keep-groups

**Reply by MagicTrevor** [reply]
I double checked the image and it never uses the USER to set www-data. Will those commands work in that case? Or is there a way to map that user to the podman user similar to root?

**Reply by MagicTrevor** [reply]
I added `--group-add=keep-groups` and I get this from `id`: `uid=82(www-data) gid=82(www-data) groups=65534(nobody),65534(nobody),0(root)`. The directory still shows as owned by root.

Trying with `-v /SOURCE:/DEST:U` results in the following: `Error: failed to chown recursively host path: lchown operation not permitted`. I tried to `sudo chown -R 100081:MyUser /SOURCE` on the server side and that didn't seem to make a difference in the container for www-data either.

**Comment by rhatdan** [comment]
Since this is not an issue, I am moving it to a discussion.

**Comment by rhatdan** [comment]
If you become root on the system and then su to the UID used within the container, can you modify the content of /mnt/mynfs/?  If this is owned by root then the permissions on the mynfs directory would need to be either world writable or group writable with the UID having membership in the group that can write to the directory.

**Reply by MagicTrevor** [reply]
I chmod'ed the mount on the system to 775 and the directory in the container appeared to be owned by www-data finally but the parent was still root:root so I re-chowned on the system itself to the 100081 user which fixed it in the container and now it can read and write! Thank you so much for the help!
